src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <regex>
+#include <vector>
+
+#include "generation_config_helper.hpp"
+#include "openvino/genai/llm_pipeline.hpp"
+#include "utils.hpp"
+
+
+namespace {
+
+struct TokenIdScore {
+    int id;
+    float score;
+
+    TokenIdScore() = default;
+    TokenIdScore(int id, float score) : id(id), score(score) {}
+
+    bool operator<(const TokenIdScore& other) const {
+        return score < other.score;
+    }
+    bool operator>(const TokenIdScore& other) const {
+        return score > other.score;
+    }
+};
+
+void sampling_softmax_inplace(TokenIdScore* first, TokenIdScore* last) {
+    float max_score = std::max_element(first, last)->score;
+    float sum = 0.f;
+    for (TokenIdScore* p = first; p != last; p++) {
+        float s = std::exp(p->score - max_score);
+        p->score = s;
+        sum += s;
+    }
+    float inv_sum = 1.f / sum;
+    for (TokenIdScore* p = first; p != last; p++) {
+        p->score *= inv_sum;
+    }
+}
+
+void sampling_top_k(TokenIdScore* first, TokenIdScore* kth, TokenIdScore* last) {
+    std::nth_element(first, kth, last, std::greater<TokenIdScore>());
+}
+
+TokenIdScore* sampling_top_p(TokenIdScore* first, TokenIdScore* last, float top_p) {

```

Comment:
```suggestion
TokenIdScore* sample_top_p(TokenIdScore* first, TokenIdScore* last, float top_p) {
```
Functions do something thus they are usually called with verbs

Reply:
done